src/ExpireNFTv4.sol

```diff
@@ -0,0 +1,114 @@
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(MINTER_ROLE, msg.sender);
+    }
+
+    function safeMint(address to) public onlyRole(MINTER_ROLE) {
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+
+        require(!_exists(tokenId), "ExpireNFTv3: invalid token ID");
+
+        _safeMint(to, tokenId);
+
+        expireTime[tokenId] = block.timestamp + EXPIRE_TIME_IN_SECONDS;  
+    }
+
+    function addOneYearOfSubscription(uint256 tokenId) public payable {
+        require(!_exists(tokenId), "ExpireNFTv4: invalid token ID");
+        require(ownerOf(tokenId) == msg.sender, "ExpireNFTv4: not the owner");
+        require(msg.value == SUBSCRIPTION_PRICE, "ExpireNFTv4: not the owner");
+
+        expireTime[tokenId] = block.timestamp + EXPIRE_TIME_IN_SECONDS;
+    }
+
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        override
+        returns (string memory)
+    {
+        _requireMinted(tokenId);
+
+        bool isExpired = block.timestamp > expireTime[tokenId];
+
+        if (isExpired) {
+            string memory json = Base64.encode(
+                bytes(
+                    string(
+                        abi.encodePacked(
+                            '{"name": "',
+                            name(),
+                            " #",
+                            Strings.toString(tokenId),
+                            '", "description": "ExpireNFTv4 collection",'
+                            '"image": "https://gateway.pinata.cloud/ipfs/QmPGMhzeYwjT1Rr4mkx5nVtuaQTFYDAQMi7f9FfVJHYpPe","attributes":[{"trait_type":"Expired","value":"true"}]}'
+                        )
+                    )
+                )
+            );
+
+            return
+                string(abi.encodePacked("data:application/json;base64,", json));
+
+        } else {
+            string memory json = Base64.encode(
+                bytes(
+                    string(
+                        abi.encodePacked(
+                            '{"name": "',
+                            name(),
+                            " #",
+                            Strings.toString(tokenId),
+                            '", "description": "ExpireNFTv4 collection",'
+                            '"image": "https://gateway.pinata.cloud/ipfs/QmaX61W1rbRquMzGjWs4CKFXtRjYhHEsNdtXGV45QZQ7nV","attributes":[{"trait_type":"Expired","value":"false"}]}'
+                        )
+                    )
+                )
+            );
+
+            return
+                string(abi.encodePacked("data:application/json;base64,", json));
+        }
+    }    
+
+    // The following functions are overrides required by Solidity.
+
+    function _beforeTokenTransfer(address from, address to, uint256 tokenId, uint256 batchSize)
+        internal
+        override(ERC721, ERC721Enumerable)
+    {
+        super._beforeTokenTransfer(from, to, tokenId, batchSize);
```

Comment:
This NFTs are transferable as they are right now. Remember that if you need them to be untransferable you need to add that functionality inside this function that you are overriding.

Reply:
Pending